src/generate/rust.rs

```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
+            }
+            put!(out, "}"); // -- end fn
+        }
+    } else {
+        put!(out, "
+            fn from_sppf(
+                parser: &'a ::gll::runtime::Parser<'i, _P, _C, I>,
+                _node: ParseNode<'i, _P>,
+                _r: traverse!(typeof(ParseNode<'i, _P>) ", rule.rule.generate_traverse_shape(false, &parse_nodes), "),
+            ) -> Self {", // -- start fn
+            name,
+            "{"); // -- start struct
+        for (field_name, paths) in &rule.fields {
+            if rule.rule.field_pathset_is_refutable(paths) {
+                put!(out, field_name, ": None");
+                for path in paths {
+                    put!(out, ".or(_r");
+                    for p in path {
+                        put!(out, " .", p);
+                    }
+                    put!(out, ")");
+                }
+                put!(out, ".map(|node| Handle {
+                    node,
+                    parser,
+                    _marker: PhantomData,
+                }),");
+            } else {
+                put!(out, field_name, ": Handle { node: _r");
+                assert_eq!(paths.len(), 1);
+                for p in paths.get_index(0).unwrap() {
+                    put!(out, " .", p);
+                }
+                put!(out, ", parser, _marker: PhantomData, },");
+            }
+        }
+
+        if rule.fields.is_empty() {
+            put!(out, "_marker: PhantomData,");
+        }
+        put!(out, "}}"); // -- end struct -- end fn
+    }
+    put!(out, "}"); // -- end impl
+
+    put!(out, "
+        impl<'a, 'i, I: ::gll::runtime::Input> Handle<'a, 'i, I, ", name, "<'a, 'i, I>> {
+            pub fn one(self) -> Result<", name, "<'a, 'i, I>, Ambiguity<Self>> {
+                // HACK(eddyb) using a closure to catch `Err`s from `?`
+                (|| Ok({
+                    #[allow(unused_variables)]
+                    let sppf = &self.parser.sppf;
+                    let node = self.node.unpack_alias();");
+    {
+        if let Some(variants) = &variants {
+            put!(out, "
+                let node = sppf.one_choice(node)?;
+                match node.kind {");
+            for Variant { rule, name: variant_name, .. } in variants {
+                put!(out, rule.parse_node_kind(&parse_nodes), " => {
+                    let r = traverse!(one(sppf, node) ", rule.generate_traverse_shape(false, &parse_nodes), ");
+                    ", name, "::", variant_name, "_from_sppf(self.parser, node, r)
+                }");
+            }
+            put!(out, "_ => unreachable!()");
+            put!(out, "}");
+        } else {
+            put!(out, "
+                let r = traverse!(one(sppf, node) ", rule.rule.generate_traverse_shape(false, &parse_nodes), ");
+                ", name, "::from_sppf(self.parser, node, r)");
+        }
+    }
+    put!(out, "
+        }))().map_err(|::gll::runtime::MoreThanOne| Ambiguity(self))
+        }
```

Comment:
Yeah, indentation was screwed up during the move - it matters much less nowadays, as we're letting `rustfmt` clean up the output (if available).

Maybe we should just remove `#[cfg_attr(rustfmt, rustfmt_skip)]` and/or switch to quasi-quotation where possible (unbalanced `{}`/`[]`/`()` being the only issue with just using `stringify!`).